### ISSUE TITLE
Switch the default driver for export_png to chromedriver

### ISSUE THREAD

**Issue by p-himik** [bokeh/bokeh#8176]
Started at #5714. All relevant details are repeated here.

PhantomJS is no longer maintained, and at this point using it produces PNGs like these:
(a fragment with y-axis)
![image](https://user-images.githubusercontent.com/4410314/44352070-8d428d00-a4cd-11e8-80be-11f5ea09208c.png)
(a fragment with x-axis with rotated labels)
![image](https://user-images.githubusercontent.com/4410314/44352074-916eaa80-a4cd-11e8-9cf1-a4259bfd5a09.png)
Using normal font instead of the italic one does not mitigate the issue.

Currently, I'm using this code to create a driver instance and pass it down to `export_png`:
```python
import os
import shutil

from chromedriver_binary.utils import get_chromedriver_path
from selenium.webdriver import Chrome, ChromeOptions

options = ChromeOptions()
# TODO: Remove when https://bugs.chromium.org/p/chromium/issues/detail?id=875241 is fixed
options.binary_location = shutil.which('chrome')

options.add_argument('--headless')
options.add_argument('--disable-gpu')
options.add_argument("--no-sandbox")
options.add_argument("--window-size=2000x2000")

web_driver = Chrome(executable_path=os.path.join(get_chromedriver_path(), 'chromedriver'),
                    chrome_options=options)
```

`chromedriver_binary` is available by installing `python-chromedriver-binary` from the `conda-forge` channel. I tested my code with version 2.41.0.

There's a few of issues:
1. Google Chrome can be available under names other than `chrome`
2. There's no check for Google Chrome version
3. The fixed window size, 2000x2000. It probably could be somehow computed in advance
4. Need to check whether this could be used with Chromium - many people use it instead of Google Chrome

Also, maybe at some point there could be a simple option that would use Firefox, but right now it appears to be unusable as it complains about missing `log` command.

**Comment by bryevdv** [comment]
Here is a complete example that sets the device pixel ratio to 1.0 to get a correctly scaled image on OSX:
```
from selenium.webdriver import Chrome, ChromeOptions

options = ChromeOptions()

options.add_argument('--headless')
options.add_argument("--window-size=2000x2000")
metrics = { "deviceMetrics": { "pixelRatio": 1.0 } }
options.add_experimental_option("mobileEmulation", metrics)
web_driver = Chrome(chrome_options=options)

import numpy as np

from bokeh.plotting import figure, show, output_file

N = 4000
x = np.random.random(size=N) * 100
y = np.random.random(size=N) * 100
radii = np.random.random(size=N) * 1.5
colors = [
    "#%02x%02x%02x" % (int(r), int(g), 150) for r, g in zip(50+2*x, 30+2*y)
]

p = figure(tools="")

p.circle(x, y, radius=radii,
          fill_color=colors, fill_alpha=0.6,
          line_color=None)

from bokeh.io import export_png

export_png(p, "out.png", webdriver=web_driver)
```
![out](https://user-images.githubusercontent.com/1078448/45395705-34f23d00-b5ec-11e8-84c1-0a05bf28d534.png)

It's unclear to me yet if just unconditionally setting to 1.0 is always the correct thing to do, or if it is a specific action for a macbook w retina display.

**Comment by RunOrVeith** [comment]
Any progress on this?

**Comment by bryevdv** [comment]
> Any progress on this?

Nope, if there were it would be described here. Given that there is a serviceable way for users to do this on their own, it has not been prioritized relative to other things. Also some investigation is needed, last time I checked, I seem to recall the Chome driver performance was a lot worse. if you re interested in working on the task we are happy to advise.

**Comment by rmelotte** [comment]
Firefox's webdriver does not currently support logs: https://github.com/SeleniumHQ/selenium/issues/2972
Do you think it would be acceptable to support Firefox's webdriver by disabling the logs when this driver is used ?
After that, exporting with Firefox is quite straightforward:
```
from selenium.webdriver import Firefox, FirefoxOptions
options = FirefoxOptions()
options.headless = True
return Firefox(options=options)
```
And the result looks ok:
![2019-04-16_15-22-41_centerRefDepth-walldistance](https://user-images.githubusercontent.com/20061471/56215332-c7b1f900-605f-11e9-892a-ddd844ca887d.png)


The following patch disables the logs when using Firefox:
```
From 7d2f49a9fd9e2fb76c8cbf8a0bb4a3438c8bc565 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Rapha=C3=ABl=20M=C3=A9lotte?= <raphael.melotte@posteo.net>
Date: Tue, 16 Apr 2019 15:40:47 +0200
Subject: [PATCH] Fixed exports when using firefox webdriver.

---
 bokeh/io/export.py | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/bokeh/io/export.py b/bokeh/io/export.py
index 4dbb53409..7a05aa380 100644
--- a/bokeh/io/export.py
+++ b/bokeh/io/export.py
@@ -267,6 +267,7 @@ def wait_until_render_complete(driver):
     '''
     from selenium.webdriver.support.ui import WebDriverWait
     from selenium.common.exceptions import TimeoutException
+    from selenium.webdriver import Firefox
 
     def is_bokeh_loaded(driver):
         return driver.execute_script('''
@@ -291,10 +292,11 @@ def wait_until_render_complete(driver):
                      a 'bokeh:idle' event to signify that the layout has rendered. \
                      Something may have gone wrong.")
     finally:
-        browser_logs = driver.get_log('browser')
-        severe_errors = [l for l in browser_logs if l.get('level') == 'SEVERE']
-        if len(severe_errors) > 0:
-            log.warning("There were severe browser errors that may have affected your export: {}".format(severe_errors))
+        if not isinstance(driver, Firefox):
+            browser_logs = driver.get_log('browser')
+            severe_errors = [l for l in browser_logs if l.get('level') == 'SEVERE']
+            if len(severe_errors) > 0:
+                log.warning("There were severe browser errors that may have affected your export: {}".format(severe_errors))
 
 #-----------------------------------------------------------------------------
 # Private API
-- 
2.11.0
```

**Comment by bryevdv** [comment]
@rmelotte if you would like to submit a PR with that change we'd be happy to accept it

**Comment by rmelotte** [comment]
For reference, the trick with `executable_path=os.path.join(get_chromedriver_path(), 'chromedriver')` doesn't work on Windows (because the binary is called "chromedriver.exe").
There is the function `chromedriver_binary.chromedriver_filename` that should point to the binary path with it's filename, but the binary doesn't seem to be there on Windows.

However, as mentioned [here](https://pypi.org/project/chromedriver-binary/), importing `chromedriver_binary` is enough to get it added to $PATH.

That said, you still have to find the Chrome executable itself, which you can do with shutil.which or with `find_executable` from `distutils.spawn`.
 In case it can help anyone, you could do something like this for it to work on Windows and Linux at the same time (I couldn't test it on OSX):
```
# TODO: Can be removed when https://bugs.chromium.org/p/chromium/issues/detail?id=875241 is fixed
from distutils.spawn import find_executable
chrome_binary_location = (
    find_executable("chrome")
    or find_executable("chromium")
    or find_executable("chrome.exe")
    or find_executable("chromium.exe")
    # hacky fallback for windows in case chrome is not in $PATH:
    or find_executable("C:\Program Files (x86)\Google\Chrome\Application\chrome.exe")
)
import chromedriver_binary
from selenium.webdriver import Chrome, ChromeOptions

options = ChromeOptions()
options.binary_location = chrome_binary_location
options.add_argument("--headless")
options.add_argument("--disable-gpu")
options.add_argument("--no-sandbox")
options.add_argument("--window-size=2000x2000")

webdriver = Chrome(options=options)
```

**Comment by bryevdv** [comment]
That code seems to work on OSX once the chromedriver package is installed. @rmelotte if you have the bandwidth to make any small PRs around these issues, it would be super helpful. If we can make it much easier to use FF and Chrome drivers, and minimally document them, that would be a great start. I think before switching the default we do need to do a performance survey. At least when I tried things last Chromedriver, was much slower than phantomjs for export.

**Comment by mattpap** [comment]
This was resolved in PR #9331.
